WordPress/Classes/ViewRelated/System/WPTabBarController.m

```diff
@@ -0,0 +1,389 @@
+#import "WPTabBarController.h"
+#import <WordPress-iOS-Shared/UIImage+Util.h>
+
+#import "WordPressAppDelegate.h"
+#import "ContextManager.h"
+#import "BlogService.h"
+#import "Blog.h"
+#import "Post.h"
+
+#import "BlogListViewController.h"
+#import "BlogDetailsViewController.h"
+#import "LoginViewController.h"
+#import "MeViewController.h"
+#import "NotificationsViewController.h"
+#import "PostsViewController.h"
+#import "ReaderPostsViewController.h"
+#import "StatsViewController.h"
+#import "WPPostViewController.h"
+#import "WPLegacyEditPageViewController.h"
+
+NSString * const WPTabBarRestorationID = @"WPTabBarID";
+NSString * const WPBlogListNavigationRestorationID = @"WPBlogListNavigationID";
+NSString * const WPReaderNavigationRestorationID= @"WPReaderNavigationID";
+NSString * const WPNotificationsNavigationRestorationID  = @"WPNotificationsNavigationID";
+
+NSString * const kWPNewPostURLParamTitleKey = @"title";
+NSString * const kWPNewPostURLParamContentKey = @"content";
+NSString * const kWPNewPostURLParamTagsKey = @"tags";
+NSString * const kWPNewPostURLParamImageKey = @"image";
+
+@interface WPTabBarController () <UITabBarControllerDelegate>
+
+@property (nonatomic, strong) BlogListViewController *blogListViewController;
+@property (nonatomic, strong) ReaderPostsViewController *readerPostsViewController;
+@property (nonatomic, strong) NotificationsViewController *notificationsViewController;
+@property (nonatomic, strong) MeViewController *meViewController;
+@property (nonatomic, strong) UIViewController *newPostViewController;
+
+@property (nonatomic, strong) UINavigationController *blogListNavigationController;
+@property (nonatomic, strong) UINavigationController *readerNavigationController;
+@property (nonatomic, strong) UINavigationController *notificationsNavigationController;
+@property (nonatomic, strong) UINavigationController *meNavigationController;
+
+@end
+
+@implementation WPTabBarController
+
++ (instancetype)sharedInstance
+{
+    static WPTabBarController *shared = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        shared = [[self alloc] init];
+    });
+    return shared;
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        self.delegate = self;
+
+        self.restorationIdentifier = WPTabBarRestorationID;
```

Comment:
Calling self.whatever is safe only for those properties that are provided by the super class. Since we're still inside of init for the subclass calling the properties defined by the subclass may inconsistent, especially if something is observing a value through KVO.